src/libraries/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/DontBeACA.cs

```diff
@@ -0,0 +1,484 @@
+
+using System.Formats.Asn1;
+using System.Net;
+using System.Net.Sockets;
+using Xunit;
+
+namespace System.Security.Cryptography.X509Certificates.Tests.CertificateCreation
+{
+    public static class DontBeACA
+    {
+        [Fact]
+        public static void EndToEnd()
+        {
+            X500DistinguishedName reqName =
+                new X500DistinguishedName("CN=Apple, CN=Banana, OU=Cherry, O=Date, L=Elderberry");
+
+            byte[] pkcs10 = BuildRequest(reqName, "Fig");
+            
+            using (X509Certificate2 issuerCert = MakeCA())
+            {
+                CertificateRequest req = IngestRequest(pkcs10, issuerCert);
+
+                byte[] serial = { 1, 1, 2, 3, 5, 8, 13, 21 };
+                DateTimeOffset notBefore = DateTimeOffset.UtcNow.AddMinutes(-2);
+                DateTimeOffset notAfter = notBefore.AddMinutes(20);
+
+                using (X509Certificate2 issued = req.Create(issuerCert, notBefore, notAfter, serial))
+                {
+                    Assert.NotNull(issued);
+                    AssertExtensions.SequenceEqual(serial, issued.SerialNumberBytes.Span);
+                    AssertExtensions.SequenceEqual(reqName.RawData, issued.SubjectName.RawData);
+                    AssertExtensions.SequenceEqual(issuerCert.SubjectName.RawData, issued.IssuerName.RawData);
+                    Assert.Equal(notBefore.DateTime, issued.NotBefore.ToUniversalTime(), TimeSpan.FromSeconds(1));
+                    Assert.Equal(notAfter.DateTime, issued.NotAfter.ToUniversalTime(), TimeSpan.FromSeconds(1));
+
+                    X509ExtensionCollection extensions = issued.Extensions;
+                    Assert.Equal(7, extensions.Count);
+
+                    // Extensions are a SEQUENCE OF, not a SET OF, so the order the CA wrote them is the order they appear.
+                    Assert.IsType<X509SubjectKeyIdentifierExtension>(extensions[0]);
+                    Assert.IsType<X509BasicConstraintsExtension>(extensions[1]);
+                    Assert.IsType<X509AuthorityKeyIdentifierExtension>(extensions[2]);
+                    Assert.Equal("2.5.29.31", extensions[3].Oid.Value);
+                    Assert.IsType<X509AuthorityInformationAccessExtension>(extensions[4]);
+                    Assert.IsType<X509KeyUsageExtension>(extensions[5]);
+                    Assert.IsType<X509SubjectAlternativeNameExtension>(extensions[6]);
+                }
+            }
+
+            static CertificateRequest IngestRequest(byte[] pkcs10, X509Certificate2 issuerCert)
+            {
+                CertificateRequest req = CertificateRequest.LoadCertificateRequest(
+                    pkcs10,
+                    HashAlgorithmName.SHA256,
+                    out int bytesConsumed,
+                    skipSignatureValidation: true,
+                    signerSignaturePadding: RSASignaturePadding.Pss);
+
+                // TODO: How to validate the Subject?
```

Comment:
That's... a terrifying TODO, since that's one of the most important things a CA does. :-)

Reply:
So, let's make X500DN have some way of enumerating the values.  Something like

`IEnumerable<(Oid NameAttributeId, string NameValue)>`.

Now you can expect that the CN (2.5.4.3) matches something acceptable, etc.  This works well for... normal... names.  But, let's imagine that the IT Department of [The Haskell Free Library and Opera House](https://en.wikipedia.org/wiki/Haskell_Free_Library_and_Opera_House) wanted to... push the boundaries of supportability and embrace their multi-location-existence.

> CN=haskellopera.com, OU=Probably an IT Department, O=Haskell Free Library and Opera House, L=Derby Line+L=Stanstead, ST=Vermont+ST=Quebec, C=US+C=CA

Or, in SEQUENCE+SET form, it looks like

* SEQUENCE OF
  * SET OF
    * C = US
    * C = CA
  * SET OF
    * ST = Vermont
    * ST = Quebec
  * SET OF
    * L = Derby Line
    * L = Stanstead
  * SET OF
    * O = Haskell Free Library and Opera House
  * SET OF
    * OU = Probably an IT Department
  * SET OF
    * CN = haskellopera.com

This level of multi-attributed RDN makes things complicated.  Should the enumerator

* throw
* Be changed to be (spiritually) `IEnumerable<IEnumerable<(Oid, string)>>`
* Be changed to something that looks a lot like an annoying union  (`IEnumerable<RelativeDistinguishedName>`, and RDN has `bool HasSimpleValue { get; }`, `Oid? SimpleAttributeId { get; }`, `string? SimpleValue { get; }`, `IEnumerable<(Oid, string)> Attributes { get; }`

The nice part about an enumerator is it makes it clear what all the parts are, just the realities of how X509Name is encoded (vs how it actually gets used) make it... complicated.  For non-"CA" reasons it might also make sense to add X500DistinguishedName.this[string]/X500DistinguishedName.this[Oid]... but that's not really sufficient for a CA approach.

Reply:
Re-wrote this today, making a `RelativeDistinguishedName` type.  Currently nested, that probably won't fly.  But it's a little cleaner.